src/ua_types.c

```diff
@@ -1090,20 +1133,57 @@ clearStructure(void *p, const UA_DataType *type) {
     for(size_t i = 0; i < type->membersSize; ++i) {
         const UA_DataTypeMember *m = &type->members[i];
         const UA_DataType *mt = &typelists[!m->namespaceZero][m->memberTypeIndex];
-        if(!m->isArray) {
-            ptr += m->padding;
-            clearJumpTable[mt->typeKind]((void*)ptr, mt);
-            ptr += mt->memSize;
+        if(m->isOptional) {
+            if((!m->isArray && (*(void* const*)ptr != NULL)) ||
+               (m->isArray && (*(void* const*)(ptr+sizeof(size_t)) != NULL))) {
+                if(!m->isArray){
+                    ptr += m->padding;
+                    UA_Array_delete(*(void**)ptr, 1, mt);
+                    ptr += sizeof(void*);
+                } else {
+                    ptr += m->padding;
+                    size_t length = *(size_t*)ptr;
+                    ptr += sizeof(size_t);
+                    UA_Array_delete(*(void**)ptr, length, mt);
+                    ptr += sizeof(void*);
+                }
+
+            } else {
```

Comment:
This code is so complicated, add some comments to branches which case they handle.